File: 1. Two Sum/shahid/two_sum.py
from typing import List
# Solution 1 (Brute Force)
    # exceeds time limit
class Solution1:
    def twoSum(self, nums: List[int], target: int) -> List[int]:

        # Looping through the array and initializing i at zeroth index
        for i in range(len(nums)-1):

            # Looping again but this time initializing j at i+1 index
            for j in range(i+1,len(nums)):
                sum = nums[i]+nums[j]

                # checking if sum is equals to target
                if sum == target:
                    return [i,j]

        return []

# Solution 2 (Using Hashmap)
class Solution2:
    def twoSum(self, nums: List[int], target: int) -> List[int]:

        # Creating Hashmap to store value as key and its index as value
        map = {}

        # Storing array value and its index in map
        for i in range(len(nums)):
            map[nums[i]] =  i
        
        # Looping and checking map for the remainder of current value subtracted from target
        for i in range(len(nums)):
            val = nums[i]
            rem = target - val

            # Getting index of the value
            if rem in map:

                # Making sure current index in not equal to index in map
                if i==map[rem]:
                    continue

                # returning current index and index of remainder value from the map
                return [i,map[rem]]

        # returning empty list if above conditions not satisfied
        return []

File: 1. Two Sum/shahid/twoSum.java
// Solution 1 (Brute Force)
class Solution1 {
    public int[] twoSum(int[] nums, int target) {
        int[] res= new int [2];

        //Looping through the array and initializing i at zeroth index
        for(int i=0;i<nums.length-1;i++){

            // Looping again but this time initializing j at i+1 index
            for(int j=i+1;j<nums.length;j++){

                //checking if sum is equals to target
                if(nums[i]+nums[j]==target){
                    res[0]=i;
                    res[1]=j;
                    return res;
                }
            }
        }
        return res;
    }
}

// Solution 2 (Using Hashmap)
class Solution2 {
    public int[] twoSum(int[] nums, int target) {
        
        //Creating Hashmap to store value as key and its index as value
        HashMap<Integer, Integer> map = new HashMap<>();
        
        //Storing array value and its index in map
        for(int i=0;i<nums.length;i++){
            map.put(nums[i],i);
        }

        // Looping and checking map for the remainder of current value subtracted from target
        for(int i=0;i<nums.length;i++){
            int n1=nums[i];
            int rem = target - n1;
            if(map.containsKey(target-nums[i])){

                //Getting index of the value
                int index = map.get(target-nums[i]);

                //Making sure current index in not equal to index in map
                if(index == i){
                    continue;
                }

                // returning current index and index of remainder value from the map
                return new int[]{i,index};
            }
        }

        //returning empty array if above conditions not satisfied
        return new int[]{};
    }
}